File: server_py/services/duplicate_service.py
# server_py/services/duplicate_service.py
"""
Intelligent fuzzy duplicate detection engine for ClaimFlow AI (Python).
Catches exact duplicates and fuzzy duplicates (different days, slight typos, altered merchant names).
"""

import re
from datetime import datetime
from typing import Dict, List, Optional, Any

BRAND_ALIASES = [
    {"canonical": "swiggy", "tokens": ["swiggy", "swigy"]},
    {"canonical": "zomato", "tokens": ["zomato", "zomato online"]},
    {"canonical": "uber", "tokens": ["uber", "uber india", "uber trip", "uber rides"]},
    {"canonical": "ola", "tokens": ["ola", "ola cabs", "ani technologies"]},
    {"canonical": "blue tokai", "tokens": ["blue tokai", "blue tokai coffee", "btcr"]},
    {"canonical": "starbucks", "tokens": ["starbucks", "tata starbucks"]},
    {"canonical": "amazon", "tokens": ["amazon", "amazon in", "cloudtail", "appario"]},
    {"canonical": "aws", "tokens": ["aws", "amazon web services", "aws emea"]},
    {"canonical": "indigo", "tokens": ["indigo", "interglobe aviation"]},
    {"canonical": "taj", "tokens": ["taj", "ihcl", "taj lands end", "taj hotel"]}
]

def normalize_string(s: str) -> str:
    if not s:
        return ""
    clean = re.sub(r"[^\w\s]", " ", s.lower())
    return re.sub(r"\s+", " ", clean).strip()

def get_canonical_merchant(name: str) -> str:
    norm = normalize_string(name)
    for alias in BRAND_ALIASES:
        if any(token in norm for token in alias["tokens"]):
            return alias["canonical"]
    return norm

def token_jaccard_similarity(s1: str, s2: str) -> float:
    tokens1 = set(w for w in normalize_string(s1).split(" ") if len(w) > 2)
    tokens2 = set(w for w in normalize_string(s2).split(" ") if len(w) > 2)
    if not tokens1 and not tokens2:
        return 1.0
    if not tokens1 or not tokens2:
        return 0.0
    intersection = len(tokens1.intersection(tokens2))
    union = len(tokens1.union(tokens2))
    return intersection / union if union > 0 else 0.0

def levenshtein_similarity(s1: str, s2: str) -> float:
    a = normalize_string(s1)
    b = normalize_string(s2)
    if a == b:
        return 1.0
    if not a or not b:
        return 0.0

    len_a = len(a)
    len_b = len(b)
    matrix = [[0] * (len_b + 1) for _ in range(len_a + 1)]

    for i in range(len_a + 1):
        matrix[i][0] = i
    for j in range(len_b + 1):
        matrix[0][j] = j

    for i in range(1, len_a + 1):
        for j in range(1, len_b + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            matrix[i][j] = min(
                matrix[i - 1][j] + 1,       # deletion
                matrix[i][j - 1] + 1,       # insertion
                matrix[i - 1][j - 1] + cost # substitution
            )

    dist = matrix[len_a][len_b]
    max_len = max(len_a, len_b)
    return 1.0 - (dist / max_len)

def get_days_difference(d1_str: str, d2_str: str) -> int:
    if not d1_str or not d2_str:
        return 999
    try:
        dt1 = datetime.strptime(d1_str[:10], "%Y-%m-%d")
        dt2 = datetime.strptime(d2_str[:10], "%Y-%m-%d")
        return abs((dt2 - dt1).days)
    except Exception:
        return 999

def check_duplicate_claim(
    candidate: Dict[str, Any],
    existing_claims: List[Dict[str, Any]],
    current_claim_id: Optional[str] = None
) -> Optional[Dict[str, Any]]:
    if not candidate or not existing_claims:
        return None

    try:
        target_amount = float(candidate.get("amount", 0))
    except (ValueError, TypeError):
        return None

    if target_amount <= 0:
        return None

    target_merchant = candidate.get("merchant", "") or ""
    target_date = candidate.get("date", "") or ""
    target_text = candidate.get("rawReceiptText") or candidate.get("description") or ""
    target_canonical = get_canonical_merchant(target_merchant)

    highest_score = 0.0
    best_match = None

    for existing in existing_claims:
        if current_claim_id and existing.get("id") == current_claim_id:
            continue
        if existing.get("status") == "rejected":
            continue

        existing_amount = float(existing.get("amount", 0))
        existing_merchant = existing.get("merchant", "") or ""
        existing_date = existing.get("date", "") or ""
        existing_text = existing.get("rawReceiptText") or existing.get("description") or ""
        existing_canonical = get_canonical_merchant(existing_merchant)

        match_score = 0.0
        reasons = []

        # 1. Amount match
        diff_ratio = abs(target_amount - existing_amount) / max(target_amount, existing_amount, 1.0)
        if target_amount == existing_amount:
            match_score += 0.40
            reasons.append(f"Identical amount (₹{target_amount:,.2f})")
        elif diff_ratio <= 0.02:
            match_score += 0.30
            reasons.append(f"Nearly identical amount (₹{target_amount} vs ₹{existing_amount})")

        # 2. Merchant match
        canonical_match = target_canonical and existing_canonical and (target_canonical == existing_canonical)
        lev = levenshtein_similarity(target_merchant, existing_merchant)
        jaccard = token_jaccard_similarity(target_merchant, existing_merchant)

        if canonical_match:
            match_score += 0.35
            reasons.append(f"Matching vendor brand ('{existing_merchant}')")
        elif lev > 0.65 or jaccard > 0.50:
            match_score += 0.25
            reasons.append(f"Similar vendor name ('{existing_merchant}')")

        # 3. Date match / proximity
        days = get_days_difference(target_date, existing_date)
        if days == 0:
            match_score += 0.25
            reasons.append(f"Same transaction date ({target_date})")
        elif days <= 3:
            match_score += 0.15
            reasons.append(f"Dates within {days} days ({target_date} vs {existing_date})")
        elif days <= 35:
            match_score += 0.10
            reasons.append(f"Transaction occurred within {days} days of previous claim")

        # 4. Text overlap bonus
        if target_text and existing_text:
            text_sim = token_jaccard_similarity(target_text, existing_text)
            if text_sim > 0.40:
                match_score += 0.15
                reasons.append(f"Significant receipt description overlap ({int(text_sim * 100)}% match)")

        final_score = min(0.99, round(match_score, 2))

        if final_score >= 0.65 and final_score > highest_score:
            highest_score = final_score
            best_match = {
                "isDuplicate": True,
                "confidence": final_score,
                "matchedClaimId": existing.get("id"),
                "matchedUser": existing.get("userName"),
                "matchedMerchant": existing.get("merchant"),
                "matchedAmount": existing.get("amount"),
                "matchedDate": existing.get("date"),
                "matchedStatus": existing.get("status"),
                "matchedRawText": existing.get("rawReceiptText") or existing.get("description"),
                "reasons": reasons,
                "reason": (
                    f"Potential duplicate ({int(final_score * 100)}% match) with Claim #{existing.get('id')} "
                    f"({existing.get('merchant')}, ₹{existing.get('amount'):,.2f}) filed on {existing.get('date')}. "
                    f"Status: {existing.get('status', '').upper()}."
                )
            }

    return best_match

File: server/services/duplicateService.js
// server/services/duplicateService.js
// Intelligent fuzzy duplicate detection engine for ClaimFlow AI

/**
 * Normalizes text for comparison: lowercases, strips special characters, removes noise words
 */
function normalizeString(str) {
  if (!str) return "";
  return str
    .toLowerCase()
    .replace(/[^\w\s]/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

/**
 * Common merchant aliases and brand normalizations
 */
const BRAND_ALIASES = [
  { canonical: "swiggy", tokens: ["swiggy", "swigy"] },
  { canonical: "zomato", tokens: ["zomato", "zomato online"] },
  { canonical: "uber", tokens: ["uber", "uber india", "uber trip", "uber rides"] },
  { canonical: "ola", tokens: ["ola", "ola cabs", "ani technologies"] },
  { canonical: "blue tokai", tokens: ["blue tokai", "blue tokai coffee", "btcr"] },
  { canonical: "starbucks", tokens: ["starbucks", "tata starbucks"] },
  { canonical: "amazon", tokens: ["amazon", "amazon in", "cloudtail", "appario"] },
  { canonical: "aws", tokens: ["aws", "amazon web services", "aws emea"] },
  { canonical: "indigo", tokens: ["indigo", "interglobe aviation"] },
  { canonical: "taj", tokens: ["taj", "ihcl", "taj lands end", "taj hotel"] }
];

function getCanonicalMerchant(name) {
  const norm = normalizeString(name);
  for (const alias of BRAND_ALIASES) {
    if (alias.tokens.some(token => norm.includes(token))) {
      return alias.canonical;
    }
  }
  return norm;
}

/**
 * Token Jaccard similarity (0 to 1)
 */
function tokenJaccardSimilarity(str1, str2) {
  const set1 = new Set(normalizeString(str1).split(" ").filter(w => w.length > 2));
  const set2 = new Set(normalizeString(str2).split(" ").filter(w => w.length > 2));
  if (set1.size === 0 && set2.size === 0) return 1.0;
  if (set1.size === 0 || set2.size === 0) return 0.0;

  let intersection = 0;
  for (const item of set1) {
    if (set2.has(item)) intersection++;
  }
  const union = new Set([...set1, ...set2]).size;
  return intersection / union;
}

/**
 * Levenshtein distance based similarity (0 to 1)
 */
function levenshteinSimilarity(s1, s2) {
  const a = normalizeString(s1);
  const b = normalizeString(s2);
  if (a === b) return 1.0;
  if (!a || !b) return 0.0;

  const matrix = [];
  for (let i = 0; i <= b.length; i++) {
    matrix[i] = [i];
  }
  for (let j = 0; j <= a.length; j++) {
    matrix[0][j] = j;
  }

  for (let i = 1; i <= b.length; i++) {
    for (let j = 1; j <= a.length; j++) {
      if (b.charAt(i - 1) === a.charAt(j - 1)) {
        matrix[i][j] = matrix[i - 1][j - 1];
      } else {
        matrix[i][j] = Math.min(
          matrix[i - 1][j - 1] + 1, // substitution
          matrix[i][j - 1] + 1,     // insertion
          matrix[i - 1][j] + 1      // deletion
        );
      }
    }
  }

  const distance = matrix[b.length][a.length];
  const maxLen = Math.max(a.length, b.length);
  return 1 - distance / maxLen;
}

/**
 * Difference in days between two date strings (YYYY-MM-DD)
 */
function getDaysDifference(dateStr1, dateStr2) {
  if (!dateStr1 || !dateStr2) return 999;
  const d1 = new Date(dateStr1);
  const d2 = new Date(dateStr2);
  if (isNaN(d1.getTime()) || isNaN(d2.getTime())) return 999;
  const diffTime = Math.abs(d2.getTime() - d1.getTime());
  return Math.round(diffTime / (1000 * 60 * 60 * 24));
}

/**
 * Check a candidate claim against all existing claims in database
 * Returns duplicate flag details or null if safe.
 */
export function checkDuplicateClaim(newClaim, existingClaims, currentClaimId = null) {
  if (!newClaim || !Array.isArray(existingClaims)) return null;

  const targetAmount = parseFloat(newClaim.amount);
  if (isNaN(targetAmount) || targetAmount <= 0) return null;

  const targetMerchant = newClaim.merchant || "";
  const targetDate = newClaim.date || "";
  const targetText = newClaim.rawReceiptText || newClaim.description || "";
  const targetCanonical = getCanonicalMerchant(targetMerchant);

  let bestMatch = null;
  let highestScore = 0;

  for (const existing of existingClaims) {
    // Skip self when updating an existing claim
    if (currentClaimId && existing.id === currentClaimId) continue;
    // Skip rejected claims from blocking new filings (or flag with info)
    if (existing.status === "rejected") continue;

    const existingAmount = parseFloat(existing.amount);
    const existingMerchant = existing.merchant || "";
    const existingDate = existing.date || "";
    const existingText = existing.rawReceiptText || existing.description || "";
    const existingCanonical = getCanonicalMerchant(existingMerchant);

    let matchScore = 0;
    const reasons = [];

    // 1. Amount comparison
    const amountDiffRatio = Math.abs(targetAmount - existingAmount) / Math.max(targetAmount, existingAmount);
    const exactAmountMatch = targetAmount === existingAmount;
    const closeAmountMatch = amountDiffRatio <= 0.02; // within 2%

    if (exactAmountMatch) {
      matchScore += 0.40;
      reasons.push(`Identical amount (₹${targetAmount.toLocaleString("en-IN")})`);
    } else if (closeAmountMatch) {
      matchScore += 0.30;
      reasons.push(`Nearly identical amount (₹${targetAmount} vs ₹${existingAmount})`);
    }

    // 2. Merchant comparison
    const canonicalMatch = targetCanonical && existingCanonical && (targetCanonical === existingCanonical);
    const merchantLev = levenshteinSimilarity(targetMerchant, existingMerchant);
    const merchantJaccard = tokenJaccardSimilarity(targetMerchant, existingMerchant);

    if (canonicalMatch) {
      matchScore += 0.35;
      reasons.push(`Matching vendor brand ('${existingMerchant}')`);
    } else if (merchantLev > 0.65 || merchantJaccard > 0.5) {
      matchScore += 0.25;
      reasons.push(`Similar vendor name ('${existingMerchant}')`);
    }

    // 3. Date proximity
    const daysDiff = getDaysDifference(targetDate, existingDate);
    if (daysDiff === 0) {
      matchScore += 0.25;
      reasons.push(`Same transaction date (${targetDate})`);
    } else if (daysDiff <= 3) {
      matchScore += 0.15;
      reasons.push(`Dates within ${daysDiff} days (${targetDate} vs ${existingDate})`);
    } else if (daysDiff <= 35) {
      // Re-filed 2-4 weeks later (classic delayed duplicate!)
      matchScore += 0.10;
      reasons.push(`Transaction occurred within ${daysDiff} days of previous claim`);
    }

    // 4. Raw text similarity bonus
    if (targetText && existingText) {
      const textSimilarity = tokenJaccardSimilarity(targetText, existingText);
      if (textSimilarity > 0.40) {
        matchScore += 0.15;
        reasons.push(`Significant receipt description overlap (${Math.round(textSimilarity * 100)}% match)`);
      }
    }

    // Cap at 0.99
    const finalScore = Math.min(0.99, Number(matchScore.toFixed(2)));

    if (finalScore >= 0.65 && finalScore > highestScore) {
      highestScore = finalScore;
      bestMatch = {
        isDuplicate: true,
        confidence: finalScore,
        matchedClaimId: existing.id,
        matchedUser: existing.userName,
        matchedMerchant: existing.merchant,
        matchedAmount: existing.amount,
        matchedDate: existing.date,
        matchedStatus: existing.status,
        matchedRawText: existing.rawReceiptText || existing.description,
        reasons: reasons,
        reason: `Potential duplicate (${Math.round(finalScore * 100)}% match) with Claim #${existing.id} (${existing.merchant}, ₹${existing.amount.toLocaleString("en-IN")}) filed on ${existing.date}. Status: ${existing.status.toUpperCase()}.`
      };
    }
  }

  return bestMatch;
}
